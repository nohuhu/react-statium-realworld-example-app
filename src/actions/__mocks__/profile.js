export const follow = jest.fn();
export const unfollow = jest.fn();
