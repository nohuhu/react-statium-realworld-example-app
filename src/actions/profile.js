export const loadProfile = async ({ state, set }, username) => {
  await set({ busy: true });

  const profile = await state.api.Profile.get(username);

  await set({ profile });

  await set({ busy: false });
};

export const follow = async ({ state, set }, username) => {
  const profile = await state.api.Profile.follow(username);

  await set({ profile });
};

export const unfollow = async ({ state, set }, username) => {
  const profile = await state.api.Profile.unfollow(username);

  await set({ profile });
};
