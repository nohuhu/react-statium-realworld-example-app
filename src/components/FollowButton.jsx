import React from 'react';
import { useStore, useState } from 'statium';

import { follow, unfollow } from '../actions/profile.js';

const FollowButton = ({ username, following }) => {
  const { dispatch } = useStore();

  // This state value is transient and local, it is only used when we are making
  // a backend request to follow or unfollow a user: the button should be disabled
  // while request is handled. Because of this very limited scope of exposure, it
  // does not make sense to declare this state value in a parent Store explicitly.
  // That said, we still want this component to be easily testable and we can
  // achieve this goal by using Statium useState() hook:
  // the first argument is the default state value (same as React.useState),
  // the second argument is the Store key for this state.
  // We import the key in the test and use it to manipulate the component.
  const [disabled, setDisabled] = useState(false, 'disableButton');

  const cls = `btn btn-sm action-btn ${following ? 'btn-secondary' : 'btn-outline-secondary'}`;
  const style = { opacity: disabled ? '0.5' : '1' };

  const onClick = async e => {
    e.preventDefault();

    await setDisabled(true);
    await dispatch(following ? unfollow : follow, username);
    await setDisabled(false);
  }

  return (
    <button className={cls} style={style} disabled={disabled} onClick={onClick}>
      <i className={following ? "ion-minus-round" : "ion-plus-round"} />
      &nbsp;
      {following ? `Unfollow ${username}` : `Follow ${username}`}
    </button>
  );
};

export default FollowButton;