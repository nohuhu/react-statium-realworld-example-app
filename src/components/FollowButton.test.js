import React from 'react';
import Store from 'statium';
import { mount } from 'test/enzyme.js';

jest.mock('../actions/profile.js');

import { follow, unfollow } from '../actions/profile.js';

import FollowButton from './FollowButton';

// We need to wrap the FollowButton component in a Store to be able to use it.
// No special requirements, just an empty Store will do.
const TestFollowButton = props => (
  <Store tag="test">
    <FollowButton {...props} />
  </Store>
);

describe("FollowButton component", () => {
  describe("rendering", () => {
    it("should render the 'follow' button when not following a user", () => {
      const tree = mount(<TestFollowButton username="foobaroo" following={false} />);

      expect(tree.find('FollowButton')).toMatchSnapshot();
    });

    it("should render the 'unfollow' button when following a user", () => {
      const tree = mount(<TestFollowButton username="foobaroo" following={true} />);

      expect(tree.find('FollowButton')).toMatchSnapshot();
    });

    it("should render the button disabled when disableButton state is true", async () => {
      // FollowButton component uses Statium.useState() hook to store its temporary
      // disabledness flag. Unlike React.useState(), we can wrap the tested component
      // in a Store and simply pass in the desired state to test it upfront, as well as
      // update it later.
      // The FollowButton is a leaf component that never has any children so the state key
      // can be named in any way convenient. If there is a possibility for key name collision,
      // Symbol keys can be used instead of strings to ensure that all state keys are unique.
      const tree = mount(
        <Store tag="disable" state={{ disableButton: true }}>
          <TestFollowButton username="foobaroo" following={true} />
        </Store>
      );

      expect(tree.find('button').prop('disabled')).toBe(true);

      // Update the state externally to make sure that the component is rendered correctly.
      // Note that we actually have _two_ Stores in the tree, and instance() method requires
      // only one node to be found. We can use Store tag prop for it.
      // Note that calling set() on a Store instance returns a Promise that is guaranteed
      // to be resolved only when the state is completely updated and child components
      // have finished rendering! We can use this combined with await keyword to write
      // readable tests that are easy to follow and comprehend.
      await tree.find('Store[tag="disable"]').instance().set({ disableButton: false });

      // Setting component state does not update the Enzyme wrapper tree, unfortunately.
      // We need to do this manually to make sure the state is reflected in the tree.
      tree.update();

      expect(tree.find('button').prop('disabled')).toBe(false);
    });
  });

  describe("behavior", () => {
    beforeEach(() => {
      follow.mockReset();
      unfollow.mockReset();
    });

    it("should dispatch follow action when not following the user and clicking on the button", async () => {
      // Note that we don't have to have an extra store for disableButton state here
      // since we don't intend to manipulate it.
      const tree = mount(<TestFollowButton username="foobaroo" following={false} />);

      tree.find('button').simulate('click');

      // Give the action enough time to fire. sleep() global is defined in test/setup.js
      await sleep(10);

      expect(follow).toHaveBeenCalled();
      expect(unfollow).not.toHaveBeenCalled();
      
      const args = follow.mock.calls[0];

      // First argument is Store API: { data, state, set, dispatch }. We don't need to test it.
      expect(args[1]).toBe('foobaroo');
    });

    it("should dispatch unfollow action when following the user and clicking on the button", async () => {
      const tree = mount(<TestFollowButton username="foobaroo" following={true} />);

      tree.find('button').simulate('click');

      await sleep(10);

      expect(unfollow).toHaveBeenCalled();
      expect(follow).not.toHaveBeenCalled();

      const args = unfollow.mock.calls[0];

      expect(args[1]).toBe('foobaroo');
    });
  });
});
