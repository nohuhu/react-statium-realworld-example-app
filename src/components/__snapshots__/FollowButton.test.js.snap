// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FollowButton component rendering should render the 'follow' button when not following a user 1`] = `
<FollowButton
  following={false}
  username="foobaroo"
>
  <button
    className="btn btn-sm action-btn btn-outline-secondary"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "opacity": "1",
      }
    }
  >
    <i
      className="ion-plus-round"
    />
     
    Follow foobaroo
  </button>
</FollowButton>
`;

exports[`FollowButton component rendering should render the 'unfollow' button when following a user 1`] = `
<FollowButton
  following={true}
  username="foobaroo"
>
  <button
    className="btn btn-sm action-btn btn-secondary"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "opacity": "1",
      }
    }
  >
    <i
      className="ion-minus-round"
    />
     
    Unfollow foobaroo
  </button>
</FollowButton>
`;
